bin/apply-ssh.bash

```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
+# This script will run an ansible playbook available in kubespray.
+# It will also install some python dependencies for kubespray in a virtual environment
+# It's not to be executed on its own but rather via `ck8s-kubespray apply-ssh <prefix>`.
+
+set -eu -o pipefail
+shopt -s globstar nullglob dotglob
+
+: "${SSH_KEY_LOCATION:?Missing SSH_KEY_LOCATION}"
+
+here="$(dirname "$(readlink -f "$0")")"
+# shellcheck source=common.bash
+source "${here}/common.bash"
+
+log_info "Running playbook authorized_key"
+pushd "${here}/../playbooks"
+
+ansible-playbook -i "${config[inventory_file]}" authorized_key.yml -b --extra-vars "ssh_key_location=$SSH_KEY_LOCATION" "${*}"
```

Comment:
Skip the `-b`, no need to be root for this, right?
```suggestion
ansible-playbook -i "${config[inventory_file]}" authorized_key.yml --extra-vars "ssh_key_location=$SSH_KEY_LOCATION" "${*}"
```

Reply:
I thought about this and I don't like it. This repo is a 95% Ansible solution, so we should avoid technology sprawl. Let's just add an Ansible variable `ssh_public_keys` which is stored the same way all the other variables are stored.

Reply:
We don't really have an established way of handling variables for ansible playbooks in this repo, since all playbooks we use are in the upstream kubespray repo. I think the easiest way of avoiding this extra environment variable would be to make this into a proper ansible role instead of just having the playbook, or do you have a different suggestion?

Reply:
We currently have `./bin/ck8s apply sc`, which does:

```
ansible-playbook -i ${config[inventory_file]} ${playbook} -b ${*}
```

If playbook was `.../authorized_key.yml`, with `ssh_public_keys` taken from `$CK8S_CONFIG_PATH/$CLUSTER`, then I would be very happy. I don't feel we need a full-blown role, just for importing `ansible.posix.authorized_key`.

Reply:
I guess what @Ajarmar is trying to say is that we so far have used *only kubespray variables* for ansible. We have not defined any new variables or even used ansible for anything other than kubespray.

If we add our own variables (which I think makes sense) we should either use a separate inventory (not so nice) or make sure we separate the variables clearly from what kubespray uses so we don't risk conflicts. I think the best way to achieve this is to explicitly read the variables from a file outside of `group_vars` like [described here](https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#defining-variables-in-included-files-and-roles). SO in the playbook we would have something like this:

```yaml
- hosts: all
  vars_files:
    - /path/to/ck8s/specific/external_vars.yml
```

Reply:
I see your point. I understand the need to avoid name clashes with upstream Kubespray. However, having a separate file feels overkill. I would simply use something like the `ck8s_` prefix and agree that those are "our" variables (so `ck8s_ssh_public_keys` ?). We are likely to add more variables for security hardening CK8s clusters in the future.

Reply:
I feel like putting these kinds of variables in their own file makes more sense. [The only common config files we have for every cloud provider](https://github.com/elastisys/compliantkubernetes-kubespray/tree/main/config/common/group_vars) are `etcd` and `k8s-cluster`. Putting a variable that only affects something locally on your computer (i.e. where you look for the SSH keys you want to apply) among the Kubernetes configuration variables seems a bit weird. If we add our own variables that actually affect the cluster directly in some way then sure, but that's not really the case here.

Reply:
I would definitely not *point* to a file with all the SSH public keys, rather *list all public SSH keys* in the configuration file. Pointing to a configuration file that is outside git is a recipe for disaster. See previous discussion that we had about SSH public keys in the `cloud-init` context: https://github.com/kubernetes-sigs/kubespray/pull/7242

Reply:
Yeah, I think having an ansible variables file in each cluster config repo sound good.
Somehting similar to
```ShellSession
# ${CK8S_CONFIG_PATH}/ssh_keys.yml
ssh_pub_keys_list:
  - ssh-rsa AAAAB3Nza1
  - ssh-rsa AAAAB3Nza2
  - ...
```